feat: security 인증 필터 및 유저 정보 조회 컨트롤러 구현

sung-silver opened the pull request:
<!--
name: Makers Auth Feature PR template
about: 구현한 기능과 변경 사항에 대해 구체적으로 작성해주세요
title: '[FEAT/{#이슈번호}] 기능 내용'
labels: ''
assignees: ''
-->

<!--
- 리뷰어 추가했나요?
- 허가자 추가했나요?
- 라벨 추가했나요?
-->

## ✨ 변경 사항 요약
- 변경사항을 요약하여 작성해주세요
- resolve #3

## 🔧 주요 구현 내용
- `JwtAuthenticationFilter` 구현
  - `Authorization` 헤더에서 JWT 추출
  - `JwtAuthenticationService.authenticate()`를 통해 토큰 검증 후 `MakersAuthentication` 반환
  - 검증된 인증 객체(`MakersAuthentication`)를 `SecurityContextHolder`에 등록
- `JwtAuthenticationExceptionFilter` 구현
  - JWT 인증 과정에서 발생한 예외를 처리
  - 인증 실패 시 일관된 에러 응답 제공
- `SecurityConfig`에 필터 체인 등록
  - `UsernamePasswordAuthenticationFilter` 앞단에 인증 필터 삽입
  - 인증 실패를 우선 처리할 수 있도록 예외 필터 추가
- 패키지 구조를 도메인형으로 변경

### 🔍 테스트 API 과정 요약
1. 클라이언트에서 Authorization 헤더에 `Bearer {accessToken}` 형식으로 토큰을 담아 api 요청
2. 인증 필터에서 토큰을 추출하고 검증 (이 과정에서 인증 서버로부터 public key를 조회하여 access token 검증에 사용)
3. 성공 시 `SecurityContext`에 사용자 인증 객체 저장
4. 이후 컨트롤러에서는 `Authentication`을 통해 인증 정보에 접근하여 userId, role 반환

### 🍀 테스트 API 응답 확인
![image](https://github.com/user-attachments/assets/91d4459d-5c9c-4917-b676-90ec67232159)

## 📝 참고 사항
- `refresh token` 재발급은 클라이언트 단에서 인증 서버로 직적 요청하기로 논의되어 구현하지 않았습니다
- keyId를 리소스 서버에서 관리하지 않도록 변경했습니다
  - 액세스 토큰을 발급할 때 [jwt 토큰에 kid를 포함할 수 있도록 인증 서버에서 변경](https://github.com/sopt-makers/sopt-auth-backend/pull/79)해주었습니다

sung-silver committed:
feat: extractClaims 메서드를 통해 userId, role을 추출할 수 있도록 JwtVerifier 로직 변경

sung-silver committed:
add: security 의존성 추가

sung-silver committed:
feat: makers authentication 구현

sung-silver committed:
feat: JwtVerifier에서 MakersAuthentication을 반환하도록 변경

sung-silver committed:
refactor: JwtVerifier -> JwtAuthenticationService로 클래스명 변경

sung-silver committed:
feat: BaseResponse 구현

sung-silver committed:
feat: ResponseUtil 구현

sung-silver committed:
feat: JwtAuthenticationFilter 구현

sung-silver committed:
feat: JwtAuthenticationExceptionFilter 구현

sung-silver committed:
feat: SecurityConfig 구현

sung-silver committed:
style: 패키지 구조 정리

sung-silver committed:
feat: user id와 role을 조회하는 test api 구현

sung-silver committed:
del: yml에서 jwk key-id 필드 삭제

JungYoonShin commented on src/main/java/sopt/makers/jwt/verifier/security/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  private String getAuthorizationToken(final HttpServletRequest request) {
+    String authorizationHeaderValue =
+        request.getHeader(HttpHeaders.AUTHORIZATION).substring(ACCESS_TOKEN_PREFIX);
+    return authorizationHeaderValue.trim();
```
p3

요기서 authorizationHeaderValue가 null이거나 "Bearer " 접두어가 없는 경우에 대한 예외처리를 진행해도 좋을 것 같아요~!

ex)
```java
    String header = request.getHeader(HttpHeaders.AUTHORIZATION);
    if (header == null || !header.startsWith("Bearer ")) {
        throw new JwtException(JwtFailure.JWT_MISSING_AUTH_HEADER);
    }
    return header.substring(ACCESS_TOKEN_PREFIX).trim();
```

sung-silver replied:
오 좋은 것 같네요 감사합니다!

JungYoonShin commented on src/main/java/sopt/makers/jwt/verifier/security/authentication/MakersAuthentication.java:
```diff
@@ -0,0 +1,66 @@
+    private final List<GrantedAuthority> authorities;
+    private boolean authenticated;
+
+    public MakersAuthentication(String userId, List<String> roles) {
+        this.userId = userId;
+        this.authorities = roles.stream()
+                .map(SimpleGrantedAuthority::new)
+                .collect(Collectors.toList());
```
p4

`authorities`가 불변 리스트라면 `.collect(Collectors.toList())`를 `toList()`로 바꿔도 좋을 것 같아요!

sung-silver replied:
`roles.stream().map(SimpleGrantedAuthority::new).toList()`의 결과가 
`List<SimpleGrantedAuthority>`이어서 타입 불일치 때문에 위와 같이 작성했는데요!
다음과 같이 수정하는건 어떨까요??

```
public MakersAuthentication(String userId, List<String> roles) {
        this.userId = userId;
        this.authorities = roles.stream()
                .map(role -> (GrantedAuthority) new SimpleGrantedAuthority(role))
                .toList();
    }
```

JungYoonShin replied:
`toList()`를 사용하려면 타입 캐스팅이 필요하군요..!! 성은님이 제안해주신 방법으로 수정하는 방안 좋습니다~~

JungYoonShin commented on src/main/java/sopt/makers/jwt/verifier/security/authentication/MakersAuthentication.java:
```diff
@@ -0,0 +1,66 @@
+        return userId;
+    }
+
+    @Override
+    public boolean isAuthenticated() {
+        return true;
+    }
```
p2

요거 무조건 `true`로 반환해도 괜찮나요!? 필터에서 인증 거치기 전에 아직 인증되지 않은 객체인데도 인증된 것처럼 동작할 수도 있지 않나 싶어서요!!

sung-silver replied:
`JwtAuthenticationService.authenticate()`가 JWT를 검증한 후에만 `MakersAuthentication`을 생성하므로,
"인증 전 객체"가 만들어질 일은 없을거라고 생각해서 true를 반환하도록 구현했는데요!

그러나 정윤님 리뷰를 보고 코드를 다시 보니 인증이 끝난 후에만 `setAuthenticated(true)`를 호출하고 있어서 코드의 일관성을 위해 true를 반환하기보다는 authenticate 필드의 상태 값을 반환하면 좋겠다는 생각이 들었습니다! 리뷰  반영할게요 감사합니다 ㅎㅎ

JungYoonShin reviewed (changes requested):
넘넘 수고하셨습니다!!! 최고최고~~

리뷰 달아두었습니다!

sung-silver committed:
[FEAT] 인증 헤더 없음에 대한 예외처리 추가

sung-silver committed:
[FEAT] MakersAuthentication에서 isAuthenticated 메서드 반환값 변경

- authenticated 필드를 false로 초기화
- isAuthenticated 메서드에서 authenticated 변수의 값을 반환하도록 변경

JungYoonShin reviewed (approved):
커밋해주시고 머지하셔도 될 것 같아요~~ 수고하셨습니다!!

sung-silver committed:
refactor: roles를 authorities로 변환하는 과정에서 toList()를 사용하도록 리팩토링
